libs/agno/agno/document/chunking/markdown.py

```diff
@@ -0,0 +1,171 @@
+import re
+from typing import List, Optional
+
+from agno.document.base import Document
+from agno.document.chunking.strategy import ChunkingStrategy
+
+
+class MarkdownChunking(ChunkingStrategy):
+    """
+    Chunking strategy optimized for Markdown documents.
+    It considers headers, paragraphs and code blocks to decide where to split the document.
+    """
+
+    def __init__(self, chunk_size: int = 5000, overlap: int = 0):
+        self.chunk_size = chunk_size
+        self.overlap = overlap
+        self._newline_size = 1
+
+    def _create_chunk_document(
+        self,
+        content: str,
+        document: Document,
+        chunk_number: int,
+        meta_data: dict,
+        current_header: Optional[str] = None,
+    ) -> Document:
+        """Create a new Document chunk with the given content and metadata."""
+        meta_data = meta_data.copy()
+        meta_data["chunk"] = chunk_number
+        meta_data["chunk_size"] = len(content)
+        if current_header:
+            meta_data["header"] = current_header
+
+        chunk_id = None
+        if document.id:
+            chunk_id = f"{document.id}_{chunk_number}"
+        elif document.name:
+            chunk_id = f"{document.name}_{chunk_number}"
+
+        return Document(
+            id=chunk_id,
+            name=document.name,
+            meta_data=meta_data,
+            content=content,
+        )
+
+    def _is_code_block_limit(self, line: str) -> bool:
+        """Check if the line marks the start or end of a code block."""
+        return line.strip().startswith("```")
+
+    def chunk(self, document: Document) -> List[Document]:
+        """Split markdown document into chunks, considering markdown structure"""
+        if not document.content or len(document.content) <= self.chunk_size:
+            return [document]
+
+        chunks: List[Document] = []
+        chunk_meta_data = document.meta_data or {}
+        chunk_number = 1
+
+        # Regex to match headers of level 2 or higher
+        header_pattern = r"^(#{2,6})\s+(.+)$"
```

Comment:
Isn't there a nice library for markdown parsing? Like `markdown` or `mistune` ?

Reply:
Tried with `markdown` but it wasn't enough. I hear you tho, this is too custom, we won't be able to fully test it. I will check for libraries again and refactor

Reply:
Updated to use the `unstructured` library while trying to keep the process as close as possible to what we do with simple txt files. Tested with different files and groups of files, seems to work fine.